### PR TITLE
Implementation of GCP Transport

### PR DESCRIPTION
**Pull request by aksrajvanshi**
@DImuthuUpe  @smarru 

We have created the GCP Transport implementation along with a basic API Gateway implementation to initiate transfers and check the status of the transfers. You need to add the path of Google Cloud Project service account credential JSON into secrets.json for Secret Service. 

Your feedback is most welcome.

### CODE REVIEW

File: api-gateway/.gitignore [hunk 1]
```diff
@@ -0,0 +1,31 @@
+HELP.md
```
Comment (DImuthuUpe) [inline api-gateway/.gitignore:1]: What is the purpose of this api-gateway directory?

File: api-gateway/.mvn/wrapper/MavenWrapperDownloader.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
+/*
```
Comment (DImuthuUpe) [inline api-gateway/.mvn/wrapper/MavenWrapperDownloader.java:1]: Is this an auto generated file? What is the purpose of this?

File: api-gateway/.mvn/wrapper/maven-wrapper.properties [hunk 3]
```diff
@@ -0,0 +1,2 @@
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.3/apache-maven-3.6.3-bin.zip
```
Comment (DImuthuUpe) [inline api-gateway/.mvn/wrapper/maven-wrapper.properties:1]: Remove this as this is not related to the PR

File: api-gateway/mvnw [hunk 4]
```diff
@@ -0,0 +1,322 @@
+#!/bin/sh
```
Comment (DImuthuUpe) [inline api-gateway/mvnw:1]: Remove this as this is not related to the PR

File: api-gateway/mvnw.cmd [hunk 5]
```diff
@@ -0,0 +1,182 @@
+@REM ----------------------------------------------------------------------------
```
Comment (DImuthuUpe) [inline api-gateway/mvnw.cmd:1]: Remove this as this is not related to the PR

File: api-gateway/pom.xml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (DImuthuUpe) [inline api-gateway/pom.xml:1]: Remove this as this is not related to the PR

File: api-gateway/src/main/java/com/example/apigateway/ApiGatewayApplication.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package com.example.apigateway;
```
Comment (DImuthuUpe) [inline api-gateway/src/main/java/com/example/apigateway/ApiGatewayApplication.java:1]: Remove this as this is not related to the PR

File: api-gateway/src/main/java/com/example/apigateway/controller.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
+package com.example.apigateway;
```
Comment (DImuthuUpe) [inline api-gateway/src/main/java/com/example/apigateway/controller.java:1]: Remove this as this is not related to the PR

File: api-gateway/src/test/java/com/example/apigateway/ApiGatewayApplicationTests.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.example.apigateway;
```
Comment (DImuthuUpe) [inline api-gateway/src/test/java/com/example/apigateway/ApiGatewayApplicationTests.java:1]: Remove this as this is not related to the PR

File: services/resource-service/server/pom.xml [hunk 10]
```diff
@@ -47,7 +47,7 @@
         <dependency>
             <groupId>org.apache.airavata</groupId>
             <artifactId>registry-api-stubs</artifactId>
-            <version>0.19-SNAPSHOT</version>
```
Comment (DImuthuUpe) [inline services/resource-service/server/pom.xml:50]: Bring 0.19-SNAPSHOT back. You should build latest mft-integration branch of airavata repo locally to get this dependency version

File: services/resource-service/server/src/main/resources/resources.json [hunk 11]
```diff
@@ -5,18 +5,18 @@
     "resourcePath": "/tmp/1mb.txt",
     "scpStorage" : {
       "storageId": "remote-ssh-storage",
-      "host": "149.165.156.124",
```
Comment (DImuthuUpe) [inline services/resource-service/server/src/main/resources/resources.json:8]: Bring original value back

File: services/resource-service/server/src/main/resources/resources.json [hunk 12]
```diff
@@ -5,18 +5,18 @@
...
+      "host": "149.165.171.139",
       "port": 22,
       "user": "root"
     }
   },
   {
     "type": "SCP",
     "resourceId":  "remote-ssh-resource2",
-    "resourcePath": "/tmp/10mb.txt",
```
Comment (DImuthuUpe) [inline services/resource-service/server/src/main/resources/resources.json:16]: Bring original value back

File: services/resource-service/server/src/main/resources/resources.json [hunk 13]
```diff
@@ -5,18 +5,18 @@
...
+    "resourcePath": "/tmp/KAMA20200408_003653_V06",
     "scpStorage" : {
       "storageId": "remote-ssh-storage",
-      "host": "149.165.156.124",
```
Comment (DImuthuUpe) [inline services/resource-service/server/src/main/resources/resources.json:19]: Bring original value back

File: services/resource-service/server/src/main/resources/resources.json [hunk 14]
```diff
@@ -29,14 +29,20 @@
   {
     "type": "S3",
     "resourceId": "s3-file",
-    "resourcePath": "new-file.txt",
```
Comment (DImuthuUpe) [inline services/resource-service/server/src/main/resources/resources.json:32]: Bring original values back

File: services/secret-service/server/pom.xml [hunk 15]
```diff
@@ -47,7 +47,7 @@
         <dependency>
             <groupId>org.apache.airavata</groupId>
             <artifactId>airavata-credential-store-stubs</artifactId>
-            <version>0.19-SNAPSHOT</version>
```
Comment (DImuthuUpe) [inline services/secret-service/server/pom.xml:50]: Bring original value back

File: services/secret-service/server/pom.xml [hunk 16]
```diff
@@ -66,7 +66,7 @@
         <dependency>
             <groupId>org.apache.airavata</groupId>
             <artifactId>airavata-data-models</artifactId>
-            <version>0.19-SNAPSHOT</version>
```
Comment (DImuthuUpe) [inline services/secret-service/server/pom.xml:69]: Bring original value back

File: services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java [hunk 17]
```diff
@@ -170,4 +170,46 @@ public boolean updateAzureSecret(AzureSecretUpdateRequest request) throws Except
...
+            JSONArray resourceList = (JSONArray) obj;
+
+            List<GCSSecret> gcsSecrets = (List<GCSSecret>) resourceList.stream()
+                    .filter(resource -> "GCS".equals(((JSONObject) resource).get("type").toString()))
+                    .map(resource -> {
+                        JSONObject r = (JSONObject) resource;
+
+                        GCSSecret gcsSecret = GCSSecret.newBuilder()
+                                .setSecretId(r.get("secretId").toString())
+                                .setJsonCredentialsFilePath(r.get("jsonCredentialsFilePath").toString()).build();
```
Comment (DImuthuUpe) [inline services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java:191]: We can't expect the JSON file path available in agent running location. This makes very hard to make the agent stateless. You should serialize json into the secret itself

File: transport/gcp-transport/mvnw [hunk 18]
```diff
@@ -0,0 +1,310 @@
+#!/bin/sh
```
Comment (DImuthuUpe) [inline transport/gcp-transport/mvnw:1]: Remove this file

File: transport/gcp-transport/mvnw.cmd [hunk 19]
```diff
@@ -0,0 +1,182 @@
+@REM ----------------------------------------------------------------------------
```
Comment (DImuthuUpe) [inline transport/gcp-transport/mvnw.cmd:1]: Remove this file

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        GoogleCredential credential = GoogleCredential.fromStream(new FileInputStream(gcsSecret.getJsonCredentialsFilePath()),transport,jsonFactory);
+        if (credential.createScopedRequired()) {
+            Collection<String> scopes = StorageScopes.all();
+            credential = credential.createScoped(scopes);
+        }
+
+        Storage storage=new Storage.Builder(transport, jsonFactory, credential).build();
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java:64]: keep spaces around =

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+        checkInitialized();
+        ResourceServiceGrpc.ResourceServiceBlockingStub resourceClient = ResourceServiceClient.buildClient(resourceServiceHost, resourceServicePort);
+        GCSResource gcsResource = resourceClient.getGCSResource(GCSResourceGetRequest.newBuilder().setResourceId(resourceId).build());
+
+        SecretServiceGrpc.SecretServiceBlockingStub secretClient = SecretServiceClient.buildClient(secretServiceHost, secretServicePort);
+        GCSSecret gcsSecret = secretClient.getGCSSecret(GCSSecretGetRequest.newBuilder().setSecretId(credentialToken).build());
+
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        GoogleCredential credential = GoogleCredential.fromStream(new FileInputStream(gcsSecret.getJsonCredentialsFilePath()),transport,jsonFactory);
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java:58]: Read the credential form a string input stream and load the json content from the secret instead of a file

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+            Collection<String> scopes = StorageScopes.all();
+            credential = credential.createScoped(scopes);
+        }
+
+        Storage storage=new Storage.Builder(transport, jsonFactory, credential).build();
+
+        ResourceMetadata metadata = new ResourceMetadata();
+        StorageObject gcsMetadata = storage.objects().get(gcsResource.getBucketName(),gcsResource.getResourcePath()).execute();
+        metadata.setResourceSize(gcsMetadata.getSize().longValue());
+        String md5Sum=String.format("%032x", new BigInteger(1, Base64.getDecoder().decode(gcsMetadata.getMd5Hash())));
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java:69]: Spaces around =

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        GoogleCredential credential = GoogleCredential.fromStream(new FileInputStream(gcsSecret.getJsonCredentialsFilePath()),transport,jsonFactory);
+        if (credential.createScopedRequired()) {
+            Collection<String> scopes = StorageScopes.all();
+            credential = credential.createScoped(scopes);
+        }
+
+        Storage storage=new Storage.Builder(transport, jsonFactory, credential).build();
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java:93]: Spaces around =

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSReceiver.java [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+        SecretServiceGrpc.SecretServiceBlockingStub secretClient = SecretServiceClient.buildClient(secretServiceHost, secretServicePort);
+        GCSSecret gcsSecret = secretClient.getGCSSecret(GCSSecretGetRequest.newBuilder().setSecretId(credentialToken).build());
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        GoogleCredential credential = GoogleCredential.fromStream(new FileInputStream(gcsSecret.getJsonCredentialsFilePath()));
+        if (credential.createScopedRequired()) {
+            Collection<String> scopes = StorageScopes.all();
+            credential = credential.createScoped(scopes);
+        }
+        storage=new Storage.Builder(transport, jsonFactory, credential).build();
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSReceiver.java:47]: Spaces around =

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        GoogleCredential credential = GoogleCredential.fromStream(new FileInputStream(gcsSecret.getJsonCredentialsFilePath()));
+        if (credential.createScopedRequired()) {
+            Collection<String> scopes = StorageScopes.all();
+            credential = credential.createScoped(scopes);
+        }
+
+        storage=new Storage.Builder(transport, jsonFactory, credential).build();
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java:54]: Spaces around =

File: api-gateway/.gitignore (hunk 1 above)
Comment (pokearu) [reply api-gateway/.gitignore:1]: In addition to the GCS transport contribution we also thought to provide a basic Api gateway. This service accepts POST calls
/transfer and /status. 
They can be used to trigger the gRPC calls in mft Apiserviceapplication to initiate transfer between provides. 
The /transfer POST call accepts sourceId,destId etc as body params to trigger the transfer.
We hoped this would help enable MFT to be treated standalone.

File: services/resource-service/server/pom.xml (hunk 10 above)
Comment (pokearu) [reply services/resource-service/server/pom.xml:50]: The highest version available in the maven repository was .17 , hence we changed this. 
Could you give us further input on how to build .19?

File: api-gateway/.gitignore (hunk 1 above)
Comment (DImuthuUpe) [reply api-gateway/.gitignore:1]: There an standard way to convert grpc to REST using grpc gateway https://github.com/grpc-ecosystem/grpc-gateway. However it is always good to look for other alternatives but it should be a separate PR. This PR is focused on GCP transport for MFT

File: services/resource-service/server/pom.xml (hunk 10 above)
Comment (DImuthuUpe) [reply services/resource-service/server/pom.xml:50]: You can clone airavata repository https://github.com/apache/airavata and checkout the mft-integration branch. Run a mvn install on that and it will generate 0.19-SNAPSHOTS

**Comment by pokearu** [comment]
We have implemented the suggested changes. 
Please check and let us know.

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+        checkInitialized();
+        ResourceServiceGrpc.ResourceServiceBlockingStub resourceClient = ResourceServiceClient.buildClient(resourceServiceHost, resourceServicePort);
+        GCSResource gcsResource = resourceClient.getGCSResource(GCSResourceGetRequest.newBuilder().setResourceId(resourceId).build());
+
+        SecretServiceGrpc.SecretServiceBlockingStub secretClient = SecretServiceClient.buildClient(secretServiceHost, secretServicePort);
+        GCSSecret gcsSecret = secretClient.getGCSSecret(GCSSecretGetRequest.newBuilder().setSecretId(credentialToken).build());
+
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        String jsonString = gcsSecret.getJsonCredentialsFilePath();
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java:62]: I don't understand what you are doing here. Do you get the file path and feed it as the file content? Is this code working?

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+        GoogleCredential credential = GoogleCredential.fromStream(new ByteArrayInputStream(jsonString.getBytes(StandardCharsets.UTF_8)), transport, jsonFactory);
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java:93]: Same as above comment

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSReceiver.java [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+    public void init(String resourceId, String credentialToken, String resourceServiceHost, int resourceServicePort, String secretServiceHost, int secretServicePort) throws Exception {
+        ResourceServiceGrpc.ResourceServiceBlockingStub resourceClient = ResourceServiceClient.buildClient(resourceServiceHost, resourceServicePort);
+        this.gcsResource = resourceClient.getGCSResource(GCSResourceGetRequest.newBuilder().setResourceId(resourceId).build());
+
+        SecretServiceGrpc.SecretServiceBlockingStub secretClient = SecretServiceClient.buildClient(secretServiceHost, secretServicePort);
+        GCSSecret gcsSecret = secretClient.getGCSSecret(GCSSecretGetRequest.newBuilder().setSecretId(credentialToken).build());
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        String jsonString = gcsSecret.getJsonCredentialsFilePath();
+        GoogleCredential credential = GoogleCredential.fromStream(new ByteArrayInputStream(jsonString.getBytes(StandardCharsets.UTF_8)));
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSReceiver.java:47]: Same as above comments

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSReceiver.java [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Override
+    public void startStream(ConnectorContext context) throws Exception {
+        logger.info("Starting GCS Receiver stream for transfer {}", context.getTransferId());
+
+        InputStream inputStream = storage.objects().get(this.gcsResource.getBucketName(), this.gcsResource.getResourcePath()).executeMediaAsInputStream();
+        OutputStream os = context.getStreamBuffer().getOutputStream();
+        int read;
+        long bytes = 0;
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSReceiver.java:67]: Instead of reading byte  by byte, use read(byte[]) method of input stream. This will improve the performance. Have a look at this example https://github.com/apache/airavata-mft/blob/master/transport/azure-transport/src/main/java/org/apache/airavata/mft/transport/azure/AzureReceiver.java#L85

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java [hunk 30]
```diff
@@ -0,0 +1,86 @@
...
+
+        ResourceServiceGrpc.ResourceServiceBlockingStub resourceClient = ResourceServiceClient.buildClient(resourceServiceHost, resourceServicePort);
+        this.gcsResource = resourceClient.getGCSResource(GCSResourceGetRequest.newBuilder().setResourceId(resourceId).build());
+
+        SecretServiceGrpc.SecretServiceBlockingStub secretClient = SecretServiceClient.buildClient(secretServiceHost, secretServicePort);
+        GCSSecret gcsSecret = secretClient.getGCSSecret(GCSSecretGetRequest.newBuilder().setSecretId(credentialToken).build());
+
+        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+        JsonFactory jsonFactory = new JacksonFactory();
+        String jsonString = gcsSecret.getJsonCredentialsFilePath();
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java:51]: Same as above comments

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+
+    }
+
+    @Override
+    public void startStream(ConnectorContext context) throws Exception {
+        logger.info("Starting GCS Sender stream for transfer {}", context.getTransferId());
+        logger.info("Content length for transfer {} {}", context.getTransferId(), context.getMetadata().getResourceSize());
+
+        InputStreamContent contentStream = new InputStreamContent(
+                "text/plain", context.getStreamBuffer().getInputStream());
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java:72]: Is the content always "text/plain"?

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java [hunk 32]
```diff
@@ -0,0 +1,86 @@
...
+        StorageObject objectMetadata = new StorageObject()
+                // Set the destination object name
+                .setName(this.gcsResource.getResourcePath())
+                // Set the access control list to publicly read-only
+                .setAcl(Arrays.asList(new ObjectAccessControl().setEntity("allUsers").setRole("READER")));
```
Comment (DImuthuUpe) [inline transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java:77]: What is this allUsers and READER property? Is this making the resource publicly available?

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java (hunk 26 above)
Comment (pokearu) [reply transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java:62]: In our implementation we get the file path from the secrets.json but we read it and store the entire json as a string in the gcsSecret. 
So we are reading the file contents in the SecretService itself and not in the agent.

I think the name in the proto message should be something other than jsonFileCredentialsFilePath, as in the implementation it stores the json as a string.

File: services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java [hunk 33]
```diff
@@ -211,4 +213,57 @@ public boolean updateAzureSecret(AzureSecretUpdateRequest request) throws Except
...
+
+            List<GCSSecret> gcsSecrets = (List<GCSSecret>) resourceList.stream()
+                    .filter(resource -> "GCS".equals(((JSONObject) resource).get("type").toString()))
+                    .map(resource -> {
+                        JSONObject r = (JSONObject) resource;
+                        StringBuilder contentBuilder = new StringBuilder();
+                        BufferedReader br = null;
+                        String jsonContents = "";
+                        try {
+                            br = new BufferedReader(new FileReader(r.get("jsonCredentialsFilePath").toString()));
```
[the remark this answers is not in the document]
Comment (pokearu) [reply services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java:234]: We read from the file path only in the SecretService. The agent no longer requires the file to be accessible, we assume the file is available with the SecretService.

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java (hunk 32 above)
Comment (pokearu) [reply transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java:77]: Yes. It makes the resource public and read only.

Comment (DImuthuUpe) [reply transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java:77]: Shouldn't that be private to the uploading user?

File: transport/gcp-transport/pom.xml [hunk 34]
```diff
@@ -0,0 +1,39 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
Comment (DImuthuUpe) [inline transport/gcp-transport/pom.xml:1]: Add apache header

File: services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java (hunk 33 above)
[the remark this answers is not in the document]
Comment (DImuthuUpe) [reply services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java:234]: Because this is the file based secret backend, it should read all credentials from the secrets.json file. You can add the credential json file content with a secret id to secrets json.  Then It will be consistent with other secret implementations

For example,
{
"type": "GCS",
"secret_id": "gcs-sample-secret",
"json": <json_content>
}

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java (hunk 26 above)
Comment (DImuthuUpe) [reply transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSMetadataCollector.java:62]: Good. Please change the property name to jsonFileCredentialsFilePath => credentialJson and refer to my comments given to file based secret backend

File: services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java (hunk 33 above)
[the remark this answers is not in the document]
Comment (pokearu) [reply services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java:234]: But with this approach a very large json has to be here as a string. Would that be fine?
 
The problem is GCS requires this entire file and it's contents, unlike other providers that can work with a single key.

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java (hunk 32 above)
Comment (pokearu) [reply transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java:77]: We thought since it's a new file, we'll by default set it as public. But yes I agree, we'll make it private.

File: services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java (hunk 33 above)
[the remark this answers is not in the document]
Comment (DImuthuUpe) [reply services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java:234]: If the backend is the File Based backend, it should read from the secrets.json. That's the idea. You can format the json to a single line and put it here. If you read the json form an external file path, it breaks the purpose of the file based backend.

[the remark this answers is not in the document]
Comment (pokearu) [reply services/secret-service/server/src/main/java/org/apache/airavata/mft/secret/server/backend/file/FileBasedSecretBackend.java:234]: Okay we will work on it.

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java (hunk 31 above)
Comment (sharanya17410) [reply transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java:72]: Could we pass the type as "null"?
GCS implementation is the only type of transport which requires content type of its stream. The only way to fetch this information involves changing the resource meta data class which would eventually affect all the other transport implementations. Can we pass null instead for GCP alone ? This would show up the transferred file content type as empty. 
image

**Comment by pokearu** [comment]
We have implemented the requested changes. Please do check and let us know. 👍

File: transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java (hunk 31 above)
Comment (DImuthuUpe) [reply transport/gcp-transport/src/main/java/org/apache/airavata/mft/transport/gcp/GCSSender.java]: I prefer null instead of text/plain because there can be binary files

File: services/resource-service/server/src/main/java/org/apache/airavata/mft/resource/server/backend/file/FileBasedResourceBackend.java [hunk 35]
```diff
@@ -90,10 +90,10 @@ public boolean deleteSCPStorage(SCPStorageDeleteRequest request) {
                         JSONObject r = (JSONObject) resource;
 
                         SCPStorage storage = SCPStorage.newBuilder()
-                                .setStorageId(((JSONObject)r.get("scpStorage")).get("storageId").toString())
-                                .setHost(((JSONObject)r.get("scpStorage")).get("host").toString())
-                                .setUser(((JSONObject)r.get("scpStorage")).get("user").toString())
-                                .setPort(Integer.parseInt(((JSONObject)r.get("scpStorage")).get("port").toString())).build();
+                                .setStorageId(((JSONObject) r.get("scpStorage")).get("storageId").toString())
```
Comment (DImuthuUpe) [inline services/resource-service/server/src/main/java/org/apache/airavata/mft/resource/server/backend/file/FileBasedResourceBackend.java:93]: Is this an intentional change? Seems like not relevant for this PR

Comment (aksrajvanshi) [reply services/resource-service/server/src/main/java/org/apache/airavata/mft/resource/server/backend/file/FileBasedResourceBackend.java:93]: Hello,

So this is one of the indentation changes produced by our IDE. Do you want us to revert it?

**Comment by pokearu** [comment]
@DImuthuUpe Okay we will add the license headers 👍

Sorry, but i am a bit unclear on what are the minor suggestions you are referring to? Is it about the indentation or the content type as null?

File: services/resource-service/server/src/main/java/org/apache/airavata/mft/resource/server/backend/file/FileBasedResourceBackend.java (hunk 35 above)
Comment (DImuthuUpe) [reply services/resource-service/server/src/main/java/org/apache/airavata/mft/resource/server/backend/file/FileBasedResourceBackend.java:93]: Yes please do. Try to keep the changes as much as relevant to the PR

**Comment by DImuthuUpe** [comment]
> @DImuthuUpe Okay we will add the license headers 👍
> 
> Sorry, but i am a bit unclear on what are the minor suggestions you are referring to? Is it about the indentation or the content type as null?

It's the indentation issue. I saw that you have changed the content type to null

**Comment by aksrajvanshi** [comment]
We have changed the indentation issue and added the license headers. Can you please look into the changes once again?

**Comment by DImuthuUpe** [comment]
All look good to me. Great work. I'm going to merge this
